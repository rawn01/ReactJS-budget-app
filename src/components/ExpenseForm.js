import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


class ExpenseForm extends React.Component {
   constructor(props) {
      super(props);
      this.state = {
         description: props.expense ? props.expense.description : '',
         note: props.expense ? props.expense.note : '',
         amount: props.expense ? props.expense.amount.toString() : '',
         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
         focused: false,
         error: ''
      };
      // console.log("waka-waka", this.state);
   }

   onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => {
         return {
            description: description
         };
      });
   };

   onNoteChange = (e) => {
      const note = e.target.value;
      this.setState(() => {
         return {
            note: note
         };
      });
   };

   onAmountChange = (e) => {
      const amount = e.target.value;
      if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState(() => {
            return {
               amount: amount
            }
         });
      }
   };

   onDateChange = (createdAt) => {
      if(createdAt) {
         this.setState(() => {
            return {
               createdAt: createdAt
            }
         });
      }
   };

   onFocusChange = ({ focused }) => {
      this.setState(() => {
         return {
            focused: focused
         };
      })
   };

   onFormSubmit = (e) => {
      e.preventDefault();

      if(!this.state.description || !this.state.amount) {
         // Set error state
         this.setState(() => {
            return {
               error: "Please provide description and amount"
            };
         })
      } else {
         // Clear the error
         this.setState(() => {
            return {
               error: ''
            }
         });
         
         this.props.onSubmit({
            desc: this.state.description,
            amount: parseFloat(this.state.amount),
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
         });
      }
   }

   render() {
      return (
         <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onFormSubmit}>
               <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange} />
               <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
               <SingleDatePicker date={this.state.createdAt} onDateChange={this.onDateChange} focused={this.state.focused} onFocusChange={this.onFocusChange} numberOfMonths={1} isOutsideRange={() => false} />
               <textarea placeholder="Add a note (optional)" value={this.state.note} onChange={this.onNoteChange}></textarea>
               <button>Add Expense</button>
            </form>
         </div>
      );
   }
};

export default ExpenseForm;