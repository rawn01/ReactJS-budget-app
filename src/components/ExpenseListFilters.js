import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
   constructor(props) {
      super(props);
      this.state = {
         focused: null
      };
   }

   onDatesChange = ({ startDate, endDate }) => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
   };

   onFocusChange = (focused) => {
      this.setState(() => {
         return {
            focused: focused
         };
      })
   };

   onTextChange = (e) => {   
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = (e) => {
      if(e.target.value === 'date') {
         this.props.sortByDate();
      } else if(e.target.value === "amount") {
         this.props.sortByAmount();
      }
   };

   render() {
      return (
         <div>
            <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />

            <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
               <option value="date" >Date</option>
               <option value="amount">Amount</option>
            </select>

            <DateRangePicker 
               startDate={this.props.filters.startDate}
               endDate={this.props.filters.endDate}
               onDatesChange={this.onDatesChange}
               focusedInput={this.state.focused}
               onFocusChange={this.onFocusChange}
               startDateId={'start'}
               endDateId={'end'}
               numberOfMonths={1}
               isOutsideRange={() => false}
               showClearDates={true}
            />
         </div>
      );
   }
}

const mapStateToProps = (state) => {
   return {
      filters: state.filters
   };
};

const matchDispatchToProps = (dispatch) => {
   return {
      setTextFilter: (text) => {
         return dispatch(setTextFilter(text));
      },
      sortByDate: () => {
         return dispatch(sortByDate());
      },
      sortByAmount: () => {
         return dispatch(sortByAmount());
      },
      setStartDate: (startDate) => {
         return dispatch(setStartDate(startDate));
      },
      setEndDate: (endDate) => {
         return dispatch(setEndDate(endDate));
      }
   };
};

const ConnectedExpenseListFilters = connect(mapStateToProps, matchDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;