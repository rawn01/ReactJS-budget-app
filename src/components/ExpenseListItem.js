import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => {
   return (
      <div>
         <Link to={`/edit/${props.id}`}>
            <h3>{props.description}</h3>
         </Link>
         <h3>{props.amount}</h3>
         <h4>- {props.createdAt}</h4>
      </div>
   );
};

export default ExpenseListItem;