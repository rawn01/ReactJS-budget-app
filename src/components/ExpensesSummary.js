import React from 'react';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
   return (
      <div>
         <h2>Viewing {props.expenseCount} expense(s) -  &#x20B9; {props.expensesTotal} </h2>
      </div>
   );
};


const mapStateToProps = (state) => {
   const visibleExpenses = selectExpenses(state.expenses, state.filters);

   return {
      expenseCount: visibleExpenses.length,
      expensesTotal: selectExpensesTotal(visibleExpenses)
   }
};

const connectExpensesSummary = connect(mapStateToProps)(ExpensesSummary);

export default connectExpensesSummary;