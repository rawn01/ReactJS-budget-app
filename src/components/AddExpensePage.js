import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export const AddExpensePage = (props) => {
   return (
      <div>
         <h1>Add Expense</h1>
         <ExpenseForm onSubmit={(expense) => {
            // props.dispatch(addExpense(expense));
            props.startAddExpense(expense);
            props.history.push('/');
         }} />
      </div>
   );
};

// Used for Unit-TESTING
const mapDispatchToProps = (dispatch) => {
   return {
      startAddExpense: (expense) => {
         return dispatch(startAddExpense(expense));
      }
   }
};

const ConnectedAddExpensePage = connect(undefined, mapDispatchToProps)(AddExpensePage);

export default ConnectedAddExpensePage;