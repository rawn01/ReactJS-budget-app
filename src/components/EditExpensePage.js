import React from 'react';
import { connect } from 'react-redux';

import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const EditPage = (props) => {
   return (
      <div>
         <ExpenseForm expense={props.expense} onSubmit={(expense) => {
            props.editExpense(props.expense.id, expense);
            props.history.push('/');
         }} />
         <button onClick={() => {
            props.removeExpense({ id: props.expense.id });
            props.history.push('/');
         }}>Remove</button>
      </div>
   );   
};

const mapStateToProps = (state, props) => {
   return {
      expense: state.expenses.find((expense) => {
         return props.match.params.id === expense.id;
      })
   }
};

// Used for TESTING
const mapDispatchToProps = (dispatch) => {
   return {
      editExpense: (id, expense) => {
         return dispatch(editExpense(id, expense));
      },
      removeExpense: (data) => {
         return dispatch(removeExpense(data));
      }
   };
};

const ConnectedEditPage = connect(mapStateToProps, mapDispatchToProps)(EditPage);

export default ConnectedEditPage;