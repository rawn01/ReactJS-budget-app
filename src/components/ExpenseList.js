import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
   return (
      <div>
         <h1>Expense List</h1>
         {
            props.expenses.length === 0 ? (
               <p>No expenses</p>
            ) : (
               props.expenses.map((item) => {
                  return <ExpenseListItem key={item.id} {...item} />
               })
            )
         }
      </div>
   );
};

const mapStateToProps = (state) => {
   return {
      expenses: selectExpenses(state.expenses, state.filters)
   };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);


export default ConnectedExpenseList;



// const ConnectedExpenseList = connect((state) => {
//    return {
//       expenses: state.expenses
//    };
// })(ExpenseList);


// export default ConnectedExpenseList;