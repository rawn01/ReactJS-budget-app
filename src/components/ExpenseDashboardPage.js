import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => {
   return (
      <div>
         <p>This is from ExpenseDashboardPage!</p>
         <ExpenseListFilters />
         <ExpenseList />
      </div>
   );
};

export default ExpenseDashboardPage;