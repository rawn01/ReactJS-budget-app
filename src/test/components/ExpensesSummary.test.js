import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render expensesSummary with single expense', () => {
   const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={115} />);
   expect(wrapper).toMatchSnapshot();
});

test('should correctly render expensesSummary with multiple expense', () => {
   const wrapper = shallow(<ExpensesSummary expenseCount={4} expensesTotal={1000} />);
   expect(wrapper).toMatchSnapshot();
});