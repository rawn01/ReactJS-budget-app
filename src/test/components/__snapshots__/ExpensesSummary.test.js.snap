// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expensesSummary with multiple expense 1`] = `
<div>
  <h2>
    Viewing 
    4
     expense(s) -  ₹ 
    1000
     
  </h2>
</div>
`;

exports[`should correctly render expensesSummary with single expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     expense(s) -  ₹ 
    115
     
  </h2>
</div>
`;
